jax/_src/pallas/mosaic_gpu/core.py

```diff
@@ -87,6 +87,7 @@ class GPUCompilerParams(pallas_core.CompilerParams):
   profile_space: int = 0
   profile_dir: str = ""
   thread_semantics: mgpu.core.ThreadSemantics = mgpu.core.ThreadSemantics.Lane
+  use_custom_barrier: bool = False
```

Comment:
Can you document the effect of `use_custom_barrier`?

Reply:
In Mosaic TPU the user never says if the kernel needs a custom barrier or not. It is inferred automatically by traversing the module and seeing if it ever uses the `tpu.GetBarrierSemaphoreOp` inside. That's what we should do here as well.

You might want to replace this argument with `collective_id`. See the discussion in my other comment

Reply:
The logic is here: https://github.com/jax-ml/jax/pull/26674/files#r1980106081